src/main/kotlin/eu/europa/ec/eudi/openid4vci/internal/RequestIssuanceImpl.kt

```diff
@@ -32,29 +32,61 @@ internal class RequestIssuanceImpl(
             supportedAlgorithmsAndMethods: SupportedEncryptionAlgorithmsAndMethods,
         ) {
             ensure(algorithm in supportedAlgorithmsAndMethods.algorithms) {
-                CredentialIssuanceError.ResponseEncryptionError.ResponseEncryptionAlgorithmNotSupportedByIssuer
+                ResponseEncryptionAlgorithmNotSupportedByIssuer
             }
             ensure(encryptionMethod in supportedAlgorithmsAndMethods.encryptionMethods) {
-                CredentialIssuanceError.ResponseEncryptionError.ResponseEncryptionMethodNotSupportedByIssuer
+                ResponseEncryptionMethodNotSupportedByIssuer
             }
         }
 
         when (val encryption = issuerMetadata.credentialResponseEncryption) {
-            CredentialResponseEncryption.NotSupported -> null
+            CredentialResponseEncryption.NotSupported ->
+                // Issuance server does not support Credential Response encryption.
+                // In case Wallet requires Credential Response encryption, fail.
+                when (config.credentialResponseEncryptionPolicy) {
+                    CredentialResponseEncryptionPolicy.SUPPORTED -> null
+                    CredentialResponseEncryptionPolicy.REQUIRED -> throw ResponseEncryptionRequiredByWalletButNotSupportedByIssuer
+                }
 
-            is CredentialResponseEncryption.SupportedNotRequired ->
-                if (config.preferEncryptedResponsesWhenSupported) {
-                    val supportedAlgorithmsAndMethods = encryption.encryptionAlgorithmsAndMethods
-                    responseEncryptionSpecFactory(supportedAlgorithmsAndMethods, config.keyGenerationConfig)?.apply {
+            is CredentialResponseEncryption.SupportedNotRequired -> {
+                // Issuance server supports but does not require Credential Response encryption.
+                // Fail in case Wallet requires Credential Response encryption but no crypto material can be generated,
+                // or in case algorithm/method supported by Wallet is not supported by issuance server.
+                val supportedAlgorithmsAndMethods = encryption.encryptionAlgorithmsAndMethods
+                val maybeSpec = responseEncryptionSpecFactory(supportedAlgorithmsAndMethods, config.keyGenerationConfig)
```

Comment:
Please wrap this into `runCatching{ ... }.getONull()` block.
Even the default implementation of `responseEncryptionSpecFactory` raises in some cases exception.